From: Alexander Zubkov <green@qrator.net>
Date: Sun, 31 Aug 2025 09:50:29 +0200
Subject: fix(dracut-systemd): avoid matching extra root= substrings in
 cmdline

`strstr ... 'root='` matches not only 'root' parameter, but any alike
substring. It could be other parameter ending in 'root', for example
'cryto_root=' or even parameter value. The presense of those in cmdline
could fail the boot process.
Fix it by adding a space in front of the matched string.

Fixes: https://github.com/dracut-ng/dracut-ng/issues/1595
Signed-off-by: Alexander Zubkov <green@qrator.net>
Origin: upstream, https://github.com/dracut-ng/dracut-ng/pull/1615
---
 modules.d/77dracut-systemd/rootfs-generator.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/modules.d/77dracut-systemd/rootfs-generator.sh b/modules.d/77dracut-systemd/rootfs-generator.sh
index 53077a7..0e87bbf 100755
--- a/modules.d/77dracut-systemd/rootfs-generator.sh
+++ b/modules.d/77dracut-systemd/rootfs-generator.sh
@@ -89,7 +89,7 @@ if [ "$rootok" = "1" ]; then
 
     generator_wait_for_dev "${root#block:}"
     generator_fsck_after_pre_mount "${root#block:}"
-    strstr "$(cat /proc/cmdline)" 'root=' || generator_mount_rootfs "${root#block:}" "$(getarg rootfstype=)" "$(getarg rootflags=)"
+    strstr " $(cat /proc/cmdline)" ' root=' || generator_mount_rootfs "${root#block:}" "$(getarg rootfstype=)" "$(getarg rootflags=)"
 fi
 
 exit 0
